### PR TITLE
[CIRCLE-36534] Section URL Share Button (Docs Disco)

### PR DESCRIPTION
**Pull request by nramabad**
# Description

This takes the current section link button and allows users to hover + click to copy the section URL to their clipboard. Original this included a icon change as well but the icon comes from Font Awesome and will be update in a future ticket.

# Reasons

JIRA: https://circleci.atlassian.net/browse/CIRCLE-36534?atlOrigin=eyJpIjoiZjliZjkwZGZhZjVlNDM4OTg5MDk1ODZiNmY0OTNiZTciLCJwIjoiaiJ9

# Screenshot

https://user-images.githubusercontent.com/42252054/133318905-9c4d5b73-2f62-4d04-8fd6-8808299d5938.mov
Screen Shot 2021-09-14 at 12 04 36 PM

### CODE REVIEW

File: jekyll/assets/js/main.js [hunk 1]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
...
   });
 
+  var tooltip = document.querySelector(".server-version-popover");
+
   // Give article headings direct links to anchors
   $("article h2, article h3, article h4, article h5, article h6").filter("[id]").each(function () {
     $(this).append('<a href="#' + $(this).attr("id") + '"><i class="fa fa-link"></i></a>');
   });
+
+  const showEvents = ['mouseover', 'hover', 'mouseenter', 'focus'];
```
Comment (teesloane) [inline jekyll/assets/js/main.js:268]: nit: `const` => `var` until we move to webpack+babel.

Comment (nramabad) [reply jekyll/assets/js/main.js:268]: https://github.com/circleci/circleci-docs/blob/99355294c22fe6a3d44938ef71a37d56817b7522/jekyll/assets/js/main.js#L129 what do you want to do about the existing consts?

Comment (teesloane) [reply jekyll/assets/js/main.js:268]: If you don't mind changing them that would be sweet - I introduced these I think, so I can also clean them up in another PR (I guess it hasn't been a huge problem 🤷‍♀️)

Comment (nramabad) [reply jekyll/assets/js/main.js]: done!

**Review by tshen815 — APPROVED** [review]
Looks good! Did we want to add a `AnalyticsClient.trackAction` to the copy link icon? @rsylvian?

File: jekyll/assets/js/main.js [hunk 2]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
...
+      event.preventDefault();
+      navigator.clipboard
+        .writeText(event.target.href)
+        .then(() => {
+          tooltip.innerHTML = "Copied!<div id='arrow' data-popper-arrow></div>";
+
+          document.querySelector(this.getAttribute('href')).scrollIntoView({
+            behavior: 'smooth'
+          });
+          window.history.pushState({}, document.title, event.target.href);
```
Comment (rsylvian) [inline jekyll/assets/js/main.js:301]: That's a nice touch!

File: jekyll/assets/js/main.js [hunk 3]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
...
+  var showEvents = ['mouseover', 'hover', 'mouseenter', 'focus'];
+  var hideEvents = [
+    'mouseout', 
+    'mouseleave', 
+    'blur'
+  ];
+  var clickEvents = ['click'];
+
+  var makePopper = (icon) => Object.assign(icon, {
+    show() {
```
Comment (rsylvian) [inline jekyll/assets/js/main.js:277]: Let's add a `trackAction` when the user sees the first tooltip

File: jekyll/assets/js/main.js [hunk 4]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
...
+          {name: 'offset',
+          options: {
+            offset: [0, 8],
+          },
+          },
+        ],
+      });
+    },
+    copy(event) {
+      event.preventDefault();
```
Comment (rsylvian) [inline jekyll/assets/js/main.js:292]: Let's add a `trackAction` when the user clicks to copy

**Review by ccgeller — APPROVED** [review]
Nice LGTM!

File: jekyll/assets/js/main.js [hunk 5]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
     $("nav.sidebar").toggleClass("open");
   });
 
+  var tooltip = document.querySelector(".server-version-popover");
```
Comment (rsylvian) [inline jekyll/assets/js/main.js:261]: Where is this coming from? Are we reusing something that was already there?

**Review by rsylvian — CHANGES_REQUESTED** [review]
Left a bunch of questions and couple comments about `trackAction`

File: jekyll/assets/js/main.js (hunk 5 above)
Comment (nramabad) [reply jekyll/assets/js/main.js:261]: yeah I'm reusing the popover for the top badges. just moved the line where it occurs. 

I could rename it generalize it.

File: jekyll/assets/js/main.js [hunk 6]
```diff
@@ -261,10 +258,81 @@ $( document ).ready(function() {
...
+          icon.parentElement.addEventListener(event, icon.hide);
+        });
+
+        clickEvents.forEach(event => {
+          icon.parentElement.addEventListener(event, icon.copy);
+        });
+      });
+    }
+  });
+
```
Comment (teesloane) [inline jekyll/assets/js/main.js:335]: Maybe we could group all experiments into a single folder for easier graduation / cleanup in the future?

Comment (nramabad) [reply jekyll/assets/js/main.js:335]: I think we can start doing that once we get webpack configured.

**Comment by teesloane** [comment]
Running locally, it seems that the `tooltip` objects is null, however, the link is correctly copying to my clipboard.

https://user-images.githubusercontent.com/12987958/133481451-15b81564-9a41-4750-b3cc-9772cde548a6.mp4

File: jekyll/assets/js/main.js (hunk 5 above)
Comment (rsylvian) [reply jekyll/assets/js/main.js:261]: That would be great! thank you

**Comment by nramabad** [comment]
> Running locally, it seems that the `tooltip` objects is null, however, the link is correctly copying to my clipboard.
> Kapture.2021-09-15.at.13.31.11.mp4

FIxed

File: jekyll/assets/js/main.js (hunk 5 above)
Comment (nramabad) [reply jekyll/assets/js/main.js]: done!

File: jekyll/assets/js/main.js (hunk 4 above)
Comment (nramabad) [reply jekyll/assets/js/main.js:292]: done!

File: jekyll/assets/js/main.js (hunk 3 above)
Comment (nramabad) [reply jekyll/assets/js/main.js:273]: done!

**Review by rsylvian — CHANGES_REQUESTED** [review]
Just tested locally and the UX is acting a little weird on my end:

Kapture 2021-09-15 at 19 05 18

1. Once you click, the page should not move at all
2. The carret on the second tooltip should not be on the far left but stay at the center

I also have this HTML link showing sometimes on the actual icon:

Kapture 2021-09-15 at 19 08 06

And the icon is also acting strange (which I think is something I noticed before in prod too tbf):

Kapture 2021-09-15 at 19 10 25

File: jekyll/assets/js/main.js [hunk 7]
```diff
@@ -261,10 +258,100 @@ $( document ).ready(function() {
...
+      navigator?.clipboard
+        .writeText(event.target.href)
+        .then(() => {
+          tooltip.innerHTML = "Copied!<div id='arrow' data-popper-arrow></div>";
+          setTimeout(() => {
+            try {
+              document.querySelector(this.getAttribute('href')).scrollIntoView({
+                behavior: 'smooth'
+              });
+              window.history.pushState({}, document.title, event.target.href);
```
Comment (rsylvian) [inline jekyll/assets/js/main.js:302]: On second thought, should we do a `replaceState` instead?

**Comment by rsylvian** [comment]
> addressed everything except the last comment which is on prod already. I'm currently thinking that's a symptom of using jquery to toggle the class. We might be able to just use css for it but I think the level of effort might be out of the scope of the story.

@nramabad Any chance it has something to do with the actual page markups? I can only reproduce reliably on https://circleci.com/docs/2.0/about-circleci/ 🤔

**Comment by rsylvian** [comment]
Yeah, that's what it is... H3 is taking the entire block:

Screen Shot 2021-09-16 at 4 25 25 PM

File: jekyll/assets/js/main.js (hunk 6 above)
Comment (rsylvian) [reply jekyll/assets/js/main.js:351]: @teesloane @nramabad 🔵  list please 🙏

**Review by rsylvian — COMMENTED** [review]
I don't think it was obvious in the jira story, but Andrew originally also wanted the button to be on the main title of the page. Is this possible to add?

**Comment by nramabad** [comment]
> I don't think it was obvious in the jira story, but Andrew originally also wanted the button to be on the main title of the page. Is this possible to add?

done

**Review by rsylvian — APPROVED** [review]
Looks good ✅ 

We now have an `experiments/` directory in `src-js`. Any chance we could move the `window.OptimizelyClient.getVariationName` part to it?

**Pull request merged**